artiq/frontend/artiq_ddb_template.py

```diff
@@ -117,25 +116,64 @@ def process_dio(self, rtio_offset, peripheral, num_channels=8):
                             "module": "artiq.coredevice.edge_counter",
                             "class": "EdgeCounter",
                             "arguments": {{"channel": 0x{channel:06x}}},
-                        }}
-                        """,
+                        }}""",
                              name=name[i],
                              channel=rtio_offset + next(channel))
         return next(channel)
 
+    def process_dio_spi(self, rtio_offset, peripheral):
+        channel = count(0)
+        for spi in peripheral["spi"]:
+            self.gen("""
+                        device_db["{name}"] = {{
+                            "type": "local",
+                            "module": "artiq.coredevice.spi2",
+                            "class": "SPIMaster",
+                            "arguments": {{"channel": 0x{channel:06x}}}
+                        }}""",
+                     name=self.get_name(spi.get("name", "dio_spi")),
+                     channel=rtio_offset + next(channel))
+        for ttl in peripheral.get("ttl", []):
+            ttl_class_names = {
+                "input": "TTLInOut",
+                "output": "TTLOut"
+            }
+            name = self.get_name(ttl.get("name", "ttl"))
```

Comment:
Do you want to provide "ttl" as a default here? Could end up with name collisions if there are multiple unnamed TTL lines in an SPI JSON description. I could be mis-understanding ``self.gen()`` though

Reply:
I think in this case `self.get_name` handles the automatic numbering of channels of a given type, e.g. "ttl" or "dio_spi" on line 134.

Reply:
Yes, that is true. So without a given name, the standard naming for TTL is used with automatic numbering.